tooltester/src/main/java/io/dockstore/tooltester/S3Client.java

```diff
@@ -0,0 +1,116 @@
+     * Workflows will be in a "workflow" directory whereas tools will be in a "tool" directory
+     * repository and optional toolname or workflowname must be encoded or else looking for logs of a specific tool without toolname (quay.io/dockstore/hello_world)
+     * will return logs for the other ones with toolnames (quay.io/dockstore/hello_world/thing)
+     *
+     * @param toolId TRS tool ID
+     * @return The key for s3
+     */
+    protected static String convertToolIdToPartialKey(String toolId) throws UnsupportedEncodingException {
+        if (toolId.startsWith("#workflow")) {
+            toolId = toolId.replaceFirst("#workflow", "workflow");
+        } else {
+            toolId = "tool/" + toolId;
+        }
+        String[] split = toolId.split("/");
+        if (split.length == MAX_TOOL_ID_STRING_SEGMENTS) {
+            split[TOOL_ID_REPOSITORY_INDEX] = URLEncoder
+                    .encode(split[TOOL_ID_REPOSITORY_INDEX] + "/" + split[TOOL_ID_TOOLNAME_INDEX], StandardCharsets.UTF_8.toString());
+            String[] encodedToolIdArray = Arrays.copyOf(split, split.length - 1);
+            return String.join("/", encodedToolIdArray);
+        } else {
+            return toolId;
+        }
+    }
+
+    /**
+     * Create s3 object with metadata and upload it to s3
+     * @param toolId    The TRS ToolId
+     * @param versionName   The TRS ToolVersion name
+     * @param testFilePath  The file that was tested (test1.json, Dockerfile)
+     * @param runner    The runner used to test the file (cwltool, cromwell)
+     * @param logContent    The contents of the log file from ToolTester
+     * @param startTime     The start time (seconds since epoch) when the file was tested
+     */
+    public void createObject(String toolId, String versionName, String testFilePath, String runner, String logContent, String startTime) {
+        try {
+            String key = generateKey(toolId, versionName, testFilePath, runner, startTime);
+            byte[] contentAsBytes = logContent.getBytes(StandardCharsets.UTF_8);
+            ByteArrayInputStream contentsAsStream = new ByteArrayInputStream(contentAsBytes);
+            ObjectMetadata metadata = new ObjectMetadata();
+            metadata.setContentType(MediaType.TEXT_PLAIN);
+            metadata.addUserMetadata("tool_id", toolId);
+            metadata.addUserMetadata("version_name", versionName);
+            metadata.addUserMetadata("test_file_path", testFilePath.replaceFirst("^Build ", "").replaceFirst("^Test ", ""));
+            metadata.addUserMetadata("runner", runner);
+            metadata.setContentLength(contentAsBytes.length);
+            PutObjectRequest request = new PutObjectRequest(bucketName, key, contentsAsStream, metadata);
+            s3.putObject(request);
+        } catch (UnsupportedEncodingException e) {
+            System.out.println("Could not generate S3 URL: " + e.getMessage());
```

Comment:
1. Use Logger
2. This is going to silently fail. Is that OK?

Reply:
2. Outputting a message is hardly considered silent (since this is a CLI) :smile:

Reply:
I didn't know if this was being invoked from a web service endpoint when I ran across the code.

If it is CLI, then do you want to do a System.exit(<non zero value>).? Dockstore CLI has an `exceptionMessage` function for this sort of thing...

Reply:
I prefer it to continue generating csv report even if it's not complete.